SU2_CFD/src/output/filewriter/CCGNSFileWriter.cpp

```diff
@@ -0,0 +1,305 @@
+        for(int j = 0; j < Buffer_Recv_nVertex[iProcessor]; j++){
+            index  = (iProcessor*MaxLocalVertex*fieldNames.size() +
+                                              j*fieldNames.size());
+            for (int i = 0; i < nVar; i++)
+                GlobalData[i][j+index2] = buffGlobData[index + i];
+        }
+        index2 += Buffer_Recv_nVertex[iProcessor];
+    }
+//    delete [] Buffer_Recv_nVertex;
+//    delete [] buffGlobData;
+  }
+// delete [] buffLocData;
+
+// Free memory: *Buffer_Recv_nVertex(solo master), *buffGlobData(solo master), *buffLocData
+  #ifdef HAVE_MPI
+      SU2_MPI::Barrier(MPI_COMM_WORLD);
+  #endif
+
+  if (rank == MASTER_NODE){
+  #ifdef HAVE_CGNS
+
+    /*---    Internal Variables    ---*/
+    int cgns_err, cgns_file, cgns_file1, cgns_flow, cgns_field, nbndry, NormalIndex, bc_wall, bc_far, bc_nulli, coord_index;
+    int nbases, base_number, cell_dim, phys_dim, nzones, zone_number, ncoords, nsections, i, S;
+    int bcnum, parent_flag;
+    cgsize_t isize[3][1];
+    cgsize_t start, end, ElementDataSize, NormalListSize, npnts, j, k, z, p, q, NewElementDataSize, dimensions;
+    cgsize_t *ConnectOffset;
+    cgsize_t pnts[1][2];
+    cgsize_t A[4], B[4];
+    cgsize_t *Elements = NULL;
+    cgsize_t *NewElements = NULL;
+    PointSetType_t ptset_type;
+    BCType_t bocotype;
+    DataType_t datatype;
+    GridLocation_t location;
+    ElementType_t type;
+    DataClass_t dataclass;
+
+    unsigned short iDim = 0, nDim = dataSorter->GetnDim();
+    unsigned short varStart = 2;
+    if (nDim == 3) varStart++;
+    char boconame;
+    char basename[] = "Base";
+    char zonename[] = "Zone 1";
+    char coordname[32];
+    char ElementSectionName[32];
+    unsigned long jVar;
+    unsigned short NVar, iVar, iPoint;
+    stringstream name, secname, ss;
+    string mesh_file, base_file, wall_string, far_string, null_string, str;
+    unsigned long nLocalVertex   = dataSorter->GetnPoints();
+    nElem = dataSorter->GetnElemGlobal();
+
+    isize[0][0] = (cgsize_t)nPoint_Global;        // vertex size
+    isize[1][0] = (cgsize_t)nElem;                // cell size
+    isize[2][0] = 0;                              //(cgsize_t)nSurf_Elem; boundary vertex size (zero if elements not sorted)
+
+    /*---    Open Mesh File    ---*/
+    mesh_file = config->GetMesh_FileName();
+    cgns_err = cg_open(mesh_file.c_str(), CG_MODE_READ, &cgns_file);
+    if(cgns_err) cg_error_print();
+    nbases = 1;
+    nzones = 1;
+
+    /*---      Create Solution File    ---*/
+    base_file = fileName;
+    cgns_err = cg_open(base_file.c_str(), CG_MODE_WRITE, &cgns_file1);
+    if(cgns_err) cg_error_print;
+    cell_dim = 3;
+    phys_dim = 3;
+    cgns_err = cg_base_write(cgns_file1, basename, cell_dim, phys_dim,  &base_number);
+    if(cgns_err) cg_error_print;
+    cgns_err = cg_zone_write(cgns_file1, base_number, zonename, *isize, Unstructured, &zone_number);
+
+    /*---    Write Coordinates    ---*/
```

Comment:
Instead of copying the mesh data it should be nice to have the possibility to create a link to the mesh file.
cg_goto(cgns_file1, base_number, "Zone_t", zone_number, "end");
std::string path = basename + "/" + zonename + "/GridCoordinates";
cg_link_write("GridCoordinates", mesh_file.c_str(), path.c_str());
...